pytorch_lightning/loops/base.py

```diff
@@ -0,0 +1,59 @@
+from abc import ABC, abstractmethod
+from typing import Any, Optional
+from weakref import proxy
+
+import pytorch_lightning as pl
+
+
+class Loop(ABC):
+
+    def __init__(self):
+        self.iteration_count: int = 0
+        self.trainer: Optional['pl.Trainer'] = None
+
+    def connect(self, trainer, *args, **kwargs) -> None:
+        """Connects Loop with all the necessary things like connectors and accelerators"""
+        self.trainer = proxy(trainer)
+
+    @property
+    @abstractmethod
+    def done(self) -> bool:
+        """Property indicating when loop is finished"""
+
+    @abstractmethod
+    def reset(self) -> None:
+        pass
+
+    def run(self, *args: Any, **kwargs: Any) -> Any:
+        self.reset()
+        self.on_run_start(*args, **kwargs)
+
+        while not self.done:
+            try:
+                self.on_advance_start(*args, **kwargs)
+                self.advance(*args, **kwargs)
+                self.on_advance_end()
+                self.iteration_count = self.increment_iteration(self.iteration_count)
+            except StopIteration:
+                break
+
+        return self.on_run_end()
+
+    def on_run_start(self, *args: Any, **kwargs: Any) -> None:
+        pass
+
+    def on_advance_start(self, *args: Any, **kwargs: Any) -> None:
+        pass
```

Comment:
This is because in various cases when you iterate over a sequence (e.g. a dataloader or a splitted batch for tbtt) this one get's passed to `run`. We then need to pass it to `on_run_start` to populate our internal state. Passing them also to `on_advance_start` and `advance` is mainly convenience but also means that we don't need to assing everything to our state here and can make use of python scopes and automatic garbage collection (adding it to the state would increase the ref counter on those objects and mean we explicitly have to free them)